crates/datafusion_ext/src/metrics.rs

```diff
@@ -287,6 +301,79 @@ impl RecordBatchStream for BoxedStreamAdapater {
     }
 }
 
+/// A wrapper around record batch stream to capture all inner datasource metrics
+/// as aggregate for a given execution.
+pub struct AggregateMetricsStreamAdapter<S> {
+    stream: S,
+    // Save the plan to capture aggregate metrics from.
+    plan: Arc<dyn ExecutionPlan>,
+    // Actual metrics store.
+    metrics: DataSourceMetrics,
+}
+
+impl<S: RecordBatchStream + Unpin> AggregateMetricsStreamAdapter<S> {
+    /// Create a new stream with a new set of data source metrics for the given
+    /// partition.
+    pub fn new(
+        stream: S,
+        plan: Arc<dyn ExecutionPlan>,
+        partition: usize,
+        metrics: &ExecutionPlanMetricsSet,
+    ) -> Self {
+        Self {
+            stream,
+            plan,
+            metrics: DataSourceMetrics::new(
+                partition,
+                metrics,
+                DataSourceMetricsOpts::read_write(),
+            ),
+        }
+    }
+}
+
+impl<S: RecordBatchStream + Unpin> Stream for AggregateMetricsStreamAdapter<S> {
+    type Item = Result<RecordBatch>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let poll = self.stream.poll_next_unpin(cx);
+
+        if let Poll::Ready(maybe_batch) = &poll {
+            // Accumulate the metrics on each poll since this is not the last exec,
+            // i.e., not guaranteed to run till the end.
+            //
+            // TODO: This might not be very efficient since [`AggregateMetrics`]
+            // iterates over the complete tree.
+            let agg_metrics = AggregatedMetrics::new_from_plan(self.plan.as_ref());
+
+            self.metrics
+                .bytes_read
+                .as_ref()
+                .expect("AggregateMetricsStreamAdapter should have bytes_read gauge")
+                .set(agg_metrics.bytes_read as usize);
+
+            self.metrics
+                .bytes_written
+                .as_ref()
+                .expect("AggregateMetricsStreamAdapter should have bytes_written gauge")
+                .set(agg_metrics.bytes_written.unwrap_or_default() as usize);
+
+            match &maybe_batch {
+                Some(Ok(batch)) => self.metrics.baseline.record_output(batch.num_rows()),
+                _ => self.metrics.baseline.done(),
+            };
+        }
```

Comment:
From what I'm seeing, we seem to be setting `bytes_written` and `bytes_read` on every batch, and _overwriting_ the existing values.

Shouldn't we only be doing this at the end? As in match on `Poll::Ready(None)` and nothing else? This would mean we're at the end of the stream, and thus only walk the execution plan once.

Reply:
> Shouldn't we only be doing this at the end? As in match on Poll::Ready(None) and nothing else? This would mean we're at the end of the stream, and thus only walk the execution plan once.

This won't work in cases where, for example, we have a `LimitExec` which stops execution before the stream completes.

Reply:
I'm assuming in that case we drop the stream? Since computing the aggregates and emitting telemetry don't block, could we implement `Drop` for this?

Reply:
> I'm assuming in that case we drop the stream? Since computing the aggregates and emitting telemetry don't block, could we implement Drop for this?

I had my doubts, and yeah, this doesn't work! We can't control when the stream is dropped. It's highly dependent on DF code and relying on dropping the stream *before* getting metrics is quite iffy!

Reply:
> From what I'm seeing, we seem to be setting `bytes_written` and `bytes_read` on every batch, and _overwriting_ the existing values.

Is overwriting a problem, in and of itself? he operation may not be efficient, but it's not incorrect? (we're not loosing data, the read/write counters are still correct?)

> Shouldn't we only be doing this at the end? As in match on `Poll::Ready(None)` and nothing else? This would mean we're at the end of the stream, and thus only walk the execution plan once.

This would certainly be more efficient, but relative to the other things going on here (e.g. sending the batch back to the client) even if it's wickedly slow (1ms? 2?) it's still a drop in the bucket compared to the transit time. (And this is isolated exclusively to hybrid exec, because the Adapter only wraps these executions.)

Reply:
My assumption, is that the two things that take measurable time here is traversing a tree that easily fits in memory, and mutex contention:
- I'd think that there's not a lot of chance for mutex contention. 
- the tree traversal seems like it could take some time particularly in pathological cases, but most of the time it's not going to be a problem, and I think even in the trivial case, the percentage of time spent doing this (relative to query execution is low.)
- if we think this is potentially a problem for user-observable latency, we could: 
  - record things lazily. 
  - store a single future to collect this information, overwriting the future on each batch, and only resolving it (at the end? on destruction?)
  - make sure it runs _after_ we start writing the batch to the socket. (maybe this is how this works?)

Reply:
> Is overwriting a problem, in and of itself? he operation may not be efficient, but it's not incorrect? (we're not loosing data, the read/write counters are still correct?)

It's not, I just misread the flow thinking we'll be getting the metrics just from the last batch read, and not the everything up to that point. This logic should be fine as is.

> This would certainly be more efficient, but relative to the other things going on here (e.g. sending the batch back to the client) even if it's wickedly slow (1ms? 2?) it's still a drop in the bucket compared to the transit time. (And this is isolated exclusively to hybrid exec, because the Adapter only wraps these executions.)

I agree, I'm not worried about this being a performance concern for us. If anything, we'll want to measure (as well as add tests for the expected metrics values so we can make changes here as necessary).

> I'd think that there's not a lot of chance for mutex contention.

At this point, mutexes/atomics will only accessed on a single thread.

> the tree traversal seems like it could take some time particularly in pathological cases, but most of the time it's not going to be a problem, and I think even in the trivial case, the percentage of time spent doing this (relative to query execution is low.)

Agreed.

> if we think this is potentially a problem for user-observable latency, we could:

I'd want to think about this a bit more. As the code is right now, I don't see this having an observable impact, and has the lowest chance of accidentally not recording a metric. I think we can get this in now, add tests, and then optimize later if we see a need for it.

Reply:
> the lowest chance of accidentally not recording a metric

I'm inclined to not worry about this error case: I think it's better (and preferable) to have a system that has a failure mode of undercounting, which would happen in situations where we have an error.

In the future, I imagine that we'll have to to do something clever/extra here if/when we have retriable errors in distributed execution.